### PR TITLE
Communicate with Gitpod instances

### PR DESCRIPTION
**Pull request by filiptronicek**
## Description

Make sure to set the correct storage items in order to make in-app nudges for installing the extension work. 

## Related Issue(s)
Falls under EXP-623

## How to test

Go to a Gitpod instance (preferably not gitpod.io) and check that `sessionStorage` (from the devtools console) has a `browser-extension-installed` key in it.

### CODE REVIEW

File: src/contents/button.tsx [hunk 1]
```diff
@@ -24,12 +26,26 @@ class ButtonContributionManager {
...
       if (!isSuitable) {
         this._disabled = true;
+        if (isSiteGitpod()) {
+          sessionStorage.setItem("browser-extension-installed", "true");
+          (async () => {
+            const gitpodEndpoint = await this._storage.getItem(STORAGE_KEY_ADDRESS);
+            const host = new URL(gitpodEndpoint).host;
+            const thisPageHost = window.location.host;
+            if (host === thisPageHost) {
+              sessionStorage.setItem("browser-extension-active", "true");
```
Comment (filiptronicek) [inline src/contents/button.tsx:43]: We use sessionStorage instead of localStorage, because we want it to be removed on tab close. This is the only way to detect that the browser extension is no longer installed.

I also thought about having the extension hide an element instead, but decided against it because it could lead to flashing the element for a frame or two.

File: src/button/button-contributions.ts [hunk 2]
```diff
@@ -31,6 +31,13 @@ export const isSiteSuitable = (): boolean => {
     return allowedApps.includes(appName);
 }
 
+/**
+ * Checks if the current site is a Gitpod instance.
+ */
+export const isSiteGitpod = (): boolean => {
+  return !!document.head.querySelector("meta[name=Gitpod]");
+}
+
```
Comment (svenefftinge) [inline src/button/button-contributions.ts:40]: this doesn't belong here. We are not contributing a button to Gitpod. Let's keep the gitpod dashboard integration completely separate.

File: src/contents/button.tsx [hunk 3]
```diff
@@ -24,6 +25,7 @@ class ButtonContributionManager {
   }
 
   _disabled = false;
+  _storage = new Storage();
```
Comment (svenefftinge) [inline src/contents/button.tsx:28]: why this field?

File: src/contents/button.tsx [hunk 4]
```diff
@@ -1,12 +1,13 @@
...
+import { Storage } from "@plasmohq/storage"
 import cssText from "data-text:../button/button.css"
 import { buttonContributions, type ButtonContributionParams, isSiteSuitable } from "../button/button-contributions";
 import { GitpodButton } from "../button/button";
 import { type ReactElement } from "react";
 import React from "react";
 
 export const config: PlasmoCSConfig = {
-  matches: ["<all_urls>"]
+  matches: ["https://*/*"]
```
Comment (svenefftinge) [inline src/contents/button.tsx:10]: why this change?

File: src/contents/gitpod-dashboard.ts [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+
+export const config: PlasmoCSConfig = {
+  matches: ["https://*/*"]
+}
+
+if (isSiteGitpod()) {
+  sessionStorage.setItem("browser-extension-installed", "true");
+  (async () => {
+    const storage = new Storage();
+    const gitpodEndpoint = await storage.getItem(STORAGE_KEY_ADDRESS);
```
Comment (svenefftinge) [inline src/contents/gitpod-dashboard.ts:14]: We don't need to distinguish whether the address points to this gitpod installation, because we tell users to install the extension, which is confusing if you have it installed.
 
We should instead show an overlay that allows users to set the address, but I believe we can do that in an overlay without interaction with gitpod dashboard.

Comment (filiptronicek) [reply src/contents/gitpod-dashboard.ts:14]: We can keep this separate and do it purely in-extension 👍

File: src/contents/button.tsx (hunk 4 above)
Comment (filiptronicek) [reply src/contents/button.tsx]: This should improve performance a tiny bit because we don't try to display buttons on some custom URL schemes like `urn://` or `file://`. Thinking twice about it (thanks!) reveals that the change is quite silly, because it would break http-only self-hosted installs of SCMs. I will keep the https requirement up for the Dashboard integration but revert the changes here.

Comment (filiptronicek) [reply src/contents/button.tsx]: Reverted in https://github.com/gitpod-io/browser-extension/pull/92/commits/353e371de634ec51e7791351e7d6b84b77a4475e

File: src/contents/button.tsx (hunk 3 above)
Comment (filiptronicek) [reply src/contents/button.tsx:28]: Oversight, we need more linting 😆

File: src/button/button-contributions.ts (hunk 2 above)
Comment (filiptronicek) [reply src/button/button-contributions.ts]: Dealt with it in 32360975ed64b9e833b341f738afe9784fe1ceed.

File: src/contents/gitpod-dashboard.ts (hunk 5 above)
Comment (filiptronicek) [reply src/contents/gitpod-dashboard.ts]: Adressed in https://github.com/gitpod-io/browser-extension/pull/92/commits/e63d3f243f9183f8521ef2c046be709f46f015d0

**Review by svenefftinge — APPROVED** [review]
Thank you! it looks very simple now. ❤️  Two nits:
1) I don't think we need the extra folder and file for the `isSiteGitpod` function. (I would just put that into the content script)
2) Is there a reason you declare that function as an object?

**Comment by filiptronicek** [comment]
> @svenefftinge: Is there a reason you declare that function as an object?

🤷 I generally do functions this way because you can really keep them nice and short (in this case, we could even just have it like below), but I don't have too much of a strong opinion on it. Can change it along with the other functions if you'd like to. 
```ts
const isSiteGitpod = () => !!document.head.querySelector("meta[name=Gitpod]");
```

**Comment by filiptronicek** [comment]
I'll be merging this now, think that we can unify this later if needed. For the automatic endpoint switching, I'll create a small follow-up (since some discussion is still ongoing)
